File: pb3/answer.js
// Time : 5:12
class Node {
  constructor(val, left=null, right=null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

const node = new Node('root', new Node('left', new Node('left.left')), new  Node('right'))

const serialize = (node) => {
  //console.log(JSON.stringify(node));
  return JSON.stringify(node);
}

const deserialize = (nodeString) => {
  const jsonObject = JSON.parse(nodeString);
  //console.log(jsonObject);
  return Object.assign(new Node, jsonObject)
}

console.log(deserialize(serialize(node)).left.left.val === 'left.left')

File: pb3/answer.py
#!/usr/bin/env python3
# Not timed
import pickle

class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def serialize(node: Node ) -> str:
  return pickle.dumps(node)
def deserialize(nodeString: str) -> Node:
  return pickle.loads(nodeString)

node = Node('root', Node('left', Node('left.left')), Node('right'))
assert deserialize(serialize(node)).left.left.val == 'left.left'
print(deserialize(serialize(node)).left.left.val == 'left.left')